### DISCUSSION TITLE
Source mapping, block step execution & line step execution.

### DISCUSSION THREAD

**Discussion by KaruroChori** [rysana-ai/bundown#7] [General]
I think it would be relevant to have some form of source mapping, so that errors in the execution of the generated script can be traced back to the original markdown file.

Two more *nice to have* in my opinion are block stepping and line stepping the execution of the file. This would give the user a chance to review what is going to be run, which is important if we are just taking and trusting a random file from the internet.

It would also be nice because I am integrating bundown with ollama and some self-hosted model to comment on the nature of the code which is going to execute, so that the inexperienced user can have a better-than-0 context about the operations which are going to be performed, if they are potentially dangerous or at the very least on what to expect.

Please, let me know what you think about these features and if they are desirable in bundown mainline.

**Comment by jrysana** [comment]
I like all of these ideas and I think they'd all be desirable in the Bundown mainline @KaruroChori! Thanks for suggesting.

1. **Source mapping:** admittedly I haven't even thought of this yet, but this is a brilliant idea and would make building a script or app directly through a Bundown file much nicer if we can deliver great, traceable errors over time.

2. **Block/line stepping execution:** we actually already have this vaguely planned in a lightweight way in terms of block stepping, but if we can also fit line/statement level stepping for supported languages into the feature set without too much complexity I'd be happy to do that too, I don't think I want it to be a *default* behavior especially since users can use any of the infinitely many approaches to building their own confirm/deny flows in-script (like with a TUI library imported into TS/JS blocks and conditional statements), but I also think having a flag to pretty print and confirm the execution of each block or line would be great for e.g. running scripts from a URL or third party source.

3. **Ollama / LLM apps built with Bundown:** I love this idea and would love to share examples of doing things like this with people to showcase what Bundown can do, so please let me know if you want to make such a thing public such that I could share it broadly :) we're also planning on doing something similar but with APIs instead of Ollama for our own purposes (we are an [LLM provider](https://rysana.com/ai))

**Comment by jrysana** [comment]
I'm gonna add these ideas to the roadmap, but please let me know if you have any thoughts on the specific terms/flags/etc.
